api/filter_dimension_options.go

```diff
@@ -108,141 +179,174 @@ func (api *FilterAPI) getFilterBlueprintDimensionOption(w http.ResponseWriter, r
 	}
 
 	if !dimensionFound {
-		log.Error(filters.ErrDimensionNotFound, logData)
-		setErrorCode(w, filters.ErrDimensionNotFound)
-		return
+		return filters.ErrDimensionNotFound
 	}
 
 	if !optionFound {
-		log.Error(filters.ErrOptionNotFound, logData)
-		setErrorCode(w, filters.ErrOptionNotFound)
-		return
+		return filters.ErrOptionNotFound
 	}
 
-	setJSONContentType(w)
-	w.WriteHeader(http.StatusNoContent)
-
-	log.Info("got dimension option for filter blueprint", logData)
+	return nil
 }
 
-func (api *FilterAPI) addFilterBlueprintDimensionOption(w http.ResponseWriter, r *http.Request) {
+func (api *FilterAPI) addFilterBlueprintDimensionOptionHandler(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
-	filterID := vars["filter_blueprint_id"]
-	name := vars["name"]
+	filterBlueprintID := vars["filter_blueprint_id"]
+	dimensionName := vars["name"]
 	option := vars["option"]
-	logData := log.Data{"filter_id": filterID, "dimension_name": name, "dimension_option": option}
+	logData := log.Data{"filter_blueprint_id": filterBlueprintID, "dimension_name": dimensionName, "dimension_option": option}
+
+	auditParams := common.Params{
+		"filter_blueprint_id": filterBlueprintID,
+		"dimension":           dimensionName,
+		"option":              option,
+	}
+	if auditErr := api.auditor.Record(r.Context(), addOptionAction, actionAttempted, auditParams); auditErr != nil {
+		handleAuditingFailure(r.Context(), addOptionAction, actionAttempted, w, auditErr, logData)
+		return
+	}
 
-	filterBlueprint, err := api.getFilterBlueprint(r.Context(), filterID)
+	err := api.addFilterBlueprintDimensionOption(r.Context(), filterBlueprintID, dimensionName, option)
 	if err != nil {
 		log.Error(err, logData)
+		if auditErr := api.auditor.Record(r.Context(), addOptionAction, actionUnsuccessful, auditParams); auditErr != nil {
+			handleAuditingFailure(r.Context(), addOptionAction, actionUnsuccessful, w, auditErr, logData)
+			return
+		}
 		switch err {
 		case filters.ErrFilterBlueprintNotFound:
 			setErrorCode(w, err, statusBadRequest)
+		case filters.ErrVersionNotFound:
+			setErrorCode(w, err, statusUnprocessableEntity)
 		default:
 			setErrorCode(w, err)
 		}
 		return
 	}
 
-	if filterBlueprint.State == models.SubmittedState {
-		log.Error(errForbidden, logData)
-		setErrorCode(w, errForbidden, filterBlueprint.State)
+	if auditErr := api.auditor.Record(r.Context(), addOptionAction, actionSuccessful, auditParams); auditErr != nil {
+		logAuditFailure(r.Context(), addOptionAction, actionSuccessful, auditErr, logData)
 		return
 	}
 
+	setJSONContentType(w)
```

Comment:
If audit successful  fails do you want to let it fall through to allow it to set the status and return type? In this scenario the dimension has already added so there is no harm in returning the appropriate success response.

Reply:
Yes you're right - I shouldn't have the return after the `logAuditFailure` 👍